test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs

```diff
@@ -252,6 +252,304 @@ public static TheoryData MultiAttributeCheckBoxData
+            var model = 23;
+            var modelExplorer = metadataProvider.GetModelExplorerForType(typeof(int), model);
+            var modelExpression = new ModelExpression(name: string.Empty, modelExplorer: modelExplorer);
+            var viewContext = TestableHtmlGenerator.GetViewContext(model, htmlGenerator, metadataProvider);
+            var tagHelper = new InputTagHelper(htmlGenerator)
+            {
+                For = modelExpression,
+                InputTypeName = inputTypeName,
+                Value = "24",
+                ViewContext = viewContext,
+            };
+
+            var attributes = new TagHelperAttributeList
+            {
+                { "type", inputTypeName },
+                { "value", "24" },
+            };
+
+            var context = new TagHelperContext(attributes, new Dictionary<object, object>(), "test");
+            var output = new TagHelperOutput(
+                "input",
+                new TagHelperAttributeList(),
+                getChildContentAsync: (useCachedResult, encoder) => Task.FromResult<TagHelperContent>(result: null))
+            {
+                TagMode = TagMode.SelfClosing,
+            };
+
+            // Act & Assert
+            ExceptionAssert.ThrowsArgument(
+                () => tagHelper.Process(context, output),
+                paramName: "expression",
+                exceptionMessage: expectedMessage);
+        }
+
+        public static TheoryData<string, TagHelperAttributeList> TypesAndExpectedAttributesData
+        {
+            get
+            {
+                return new TheoryData<string, TagHelperAttributeList>
+                {
+                    {
+                        "hidden", new TagHelperAttributeList
+                        {
+                            { "name",  "-expression-" },
+                            { "type", "hidden" },
+                            { "value", "False" },
+                        }
+                    },
+                    {
+                        "number", new TagHelperAttributeList
+                        {
+                            { "name",  "-expression-" },
+                            { "type", "number" },
+                            { "value", "False" },
+                        }
+                    },
+                    {
+                        "password", new TagHelperAttributeList
+                        {
+                            { "name",  "-expression-" },
+                            { "type", "password" },
+                        }
+                    },
+                    {
+                        "text", new TagHelperAttributeList
+                        {
+                            { "name",  "-expression-" },
+                            { "type", "text" },
+                            { "value", "False" },
+                        }
+                    },
+                };
+            }
+        }
+
+        [Theory]
```

Comment:
ditto about not using 1 fact, use helper methods.

Reply:
Again, the helper method would be the existing test.

However, I will split out the password test because that's the only one with truly unique expectations. That'll reduce this test's inputs to just `inputTypeName`.